content/gocourse-2019-05-melbourne.slide

```diff
@@ -744,50 +744,103 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 .caption J. D. Frazer "Illiad", User Friendly - [[https://en.wikipedia.org/wiki/User_Friendly#Crud_Puppy][Crud Puppy]]
 
 
-* Errors (1/3)
+* Errors in Go
 
-"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+- `error` a builtin type. ([[https://golang.org/src/builtin/builtin.go?s=11246:11286#L251][source: error]])
+- Check errors explicitly
+- The [[https://golang.org/pkg/errors/][`errors`]] package has functions for creating new error type.
+- Creation of new error type can be done in a few ways:
+
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
+	}
+- Or use formatted error message using `fmt.Errorf`
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
+	}
+
+* Errors in Go (...cotd)
 
-The `error` type is a built-in interface:
+- Custom error types can be created. Example from `net` package -
 
-	type error interface {
-	    Error() string
+	type Error interface {
+		error
+		Timeout() bool
+		Temporary() bool
 	}
+- Implementation of the `Error` interface here - [[https://golang.org/pkg/net/#AddrError][net/AddrError]]
+- The error values can be logged and printed. Example -
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
-		}
-		return catStore.cats[id], nil
+
+* Error Handling
+
+- Error handling: Check for the presence of an error (compare with nil).
+- Do not compare error variables. ([[https://goplay.space/#mgw9QAmeLUP][Example]])
+- Errors are values so it can be programmed. ([[https://goplay.space/#MTqqxwEdq-A][Example]])
+	g, err := Sqrt(-1); err != nil {
```

Comment:
I would favor code that compiles (or at least close to it), eg:
```
if _, err := Sqrt(-1); err != nil {
  log.Println(err)
}
```
or to show usage of the error and result for no error
```
sqrt, err := Sqrt(-1)
if err != nil {
  log.Println(err)
}
```

Reply:
Done